New packages: intellij-idea-ultimate-edition-2019.3.2

2asoft opened the pull request:
This based on the existing `intellij-idea-community-edition`.

I'm explicitly removing the JetBrains JDK and requiring JetBrains' custom JDK instead.

Hoshpak commented:
Are these redistributable? Otherwise they have to be marked as restricted.

Chocimier commented on srcpkgs/clion/template:
```diff
@@ -0,0 +1,58 @@
+# Template file for 'clion'
+_dlname="CLion"
+_pkgname="clion"
+pkgname=clion
```
Please use same case as tarball name. It is not only Void way, it also makes default update pattern work.

Chocimier commented on srcpkgs/clion/template:
```diff
@@ -0,0 +1,58 @@
+# Template file for 'clion'
+_dlname="CLion"
```
As it is used once, no point to be defined.

Chocimier commented on srcpkgs/clion/template:
```diff
@@ -0,0 +1,58 @@
+revision=1
+archs="x86_64"
+depends="virtual?java-environment giflib libXtst"
+short_desc="CLion: A smart cross-platform IDE for C and C++"
```
Remove package name and initial article.

Chocimier commented on srcpkgs/clion/template:
```diff
@@ -0,0 +1,58 @@
+depends="virtual?java-environment giflib libXtst"
+short_desc="CLion: A smart cross-platform IDE for C and C++"
+maintainer="Anton Afanasyev <anton@doubleasoftware.com>"
+license="Commercial"
```
License is not [spdx](https://spdx.org/licenses/), so need to be written as `custom:Commercial` and included in package.

2asoft replied:
Thanks!
I've put `custom:Commercial`, but I'm not sure whether I should be including it in the package.
There are about 30 license files for each of these applications, and I'm not sure it makes sense to include all?

Chocimier commented on srcpkgs/clion/template:
```diff
@@ -0,0 +1,58 @@
+short_desc="CLion: A smart cross-platform IDE for C and C++"
+maintainer="Anton Afanasyev <anton@doubleasoftware.com>"
+license="Commercial"
+homepage="https://www.jetbrains.org/"
```
Is there a program specific page?

Chocimier commented on srcpkgs/clion/template:
```diff
@@ -0,0 +1,58 @@
+	mkdir _tmp
+	# JetBrains' JRE can be installed separately as a subpackage.
+	# For now, just move it out of the way
+	mv jre64 _tmp/
```
why not just remove/not install that?

2asoft replied:
I intended to make further changes to this, to install the jre separately.
But I'll want to create a separate package instead, so this isn't needed, and I removed it -- thanks.

Chocimier commented on srcpkgs/clion/template:
```diff
@@ -0,0 +1,58 @@
+	vmkdir usr/share/doc/${_pkgname}
+	vmkdir usr/lib/${_pkgname}
+
+	mv bin lib plugins product-info.json build.txt ${DESTDIR}/usr/lib/${_pkgname}
```
Use vcopy, it creates needed directories. see Manual.md.

Chocimier commented on srcpkgs/clion/template:
```diff
@@ -0,0 +1,58 @@
+	vmkdir usr/lib/${_pkgname}
+
+	mv bin lib plugins product-info.json build.txt ${DESTDIR}/usr/lib/${_pkgname}
+	mv license ${DESTDIR}/usr/share/doc/${_pkgname}
```
`vlicense license`

Chocimier commented on srcpkgs/clion/template:
```diff
@@ -0,0 +1,58 @@
+	mv bin lib plugins product-info.json build.txt ${DESTDIR}/usr/lib/${_pkgname}
+	mv license ${DESTDIR}/usr/share/doc/${_pkgname}
+
+	rm ${DESTDIR}/usr/lib/${_pkgname}/bin/fsnotifier-arm -rf
```
maybe remove before installing?

Chocimier commented on srcpkgs/clion/template:
```diff
@@ -0,0 +1,58 @@
+		rm ${DESTDIR}/usr/lib/${_pkgname}/bin/libyjpagent-linux.so
+		rm ${DESTDIR}/usr/lib/${_pkgname}/lib/pty4j-native/linux/x86 -rf
+		;;
+		i686)
```
That won't happen because of `archs=` set before.

2asoft replied:
Amended the archs list.

Chocimier commented on srcpkgs/goland/template:
```diff
@@ -0,0 +1,58 @@
+nopie=yes
+
+do_extract() {
+	tar zxf ${XBPS_SRCDISTDIR}/${pkgname}-${version}/${_pkgname}-${version}.tar.gz --strip-components=1 -C ${wrksrc}
```
set `wrksrc="GoLand-${version}"` instead

Chocimier commented on srcpkgs/clion/INSTALL.msg:
```diff
@@ -0,0 +1,13 @@
+	lldb: Native LLVM debugger
+	gcc: GNU compiler
+	clang: LLVM compiler
+	biicode: C/C++ dependency manager
```
Please only list ones available in Void.

Chocimier commented:
Also, commit titles should be `New package: $pkgname-$version`

2asoft commented:
Thanks for the review!
I think I addressed all the concerns - this looks better now.

These are indeed not redistributable, so I marked as `restricted=yes`.

Do I need to include all the licenses with each of these anyway?
There are about 30 with each package, and it probably doesn't make sense to update them all every time, especially since these are restricted anyway?

2asoft commented:
I see the `XLINT` task complains about lack of `vlicense` calls for `custom` license.
There isn't a single license file here - there are many.
What is the correct Void approach in this case?

Chocimier commented on srcpkgs/CLion/INSTALL.msg:
```diff
@@ -0,0 +1,11 @@
+	clang: LLVM compiler
+	gtest: C++ testing
+	python: Python 3 programming language support
+	python2: Python 2 programming language support
```
python & python3

Chocimier commented on srcpkgs/CLion/template:
```diff
@@ -0,0 +1,62 @@
+	rm bin/fsnotifier-arm -rf
+	rm lib/pty4j-native/linux/ppc64le -rf
+	case $XBPS_TARGET_MACHINE in
+		x86_64)
```
Indent like in LuaJIT template.

Chocimier commented on srcpkgs/CLion/update:
```diff
@@ -0,0 +1,2 @@
+pattern="CLion-\K[\d.]+(?=\.tar)"
```
`CLion` is now pkgname so pattern is not necessary.

2asoft replied:
Thanks for pointing it out.
I tried removing the pattern, but the document the update site returns contains links for all OS, not just Linux.
`xbps-src update-check CLion` then returns something like this:
```
=> Using `/home/aasoft/projects/void_linux_packages/void-packages/hostdir/binpkgs/2asoft/jetbrains-intellij-pro' as local repository.
CLion-2019.1.2 -> CLion-2019.1.3
CLion-2019.1.2 -> CLion-2019.1.3.win
```
This probably doesn't break anything, but if JetBrains ever release non-Linux releases before the Linux releases, the above command would falsely claim that an update is available.

I think it'd be better explicitly have this pattern that is tailored to the Linux releases.

Chocimier replied:
There is `ignore=` for that, but you can also leave is as other templates do need pattern anyway.

Chocimier commented on srcpkgs/GoLand/template:
```diff
@@ -0,0 +1,55 @@
+pkgname=GoLand
+version=2019.1.1
+revision=1
+wrksrc="GoLand-${version}"
```
This is default, remove.

2asoft replied:
Assuming you mean `wrksrc="GoLand-${version}"` here and not any of the other highlighted variables.

Chocimier replied:
Yes, I meant default of wrksrc is pkgname-version.

Chocimier commented on srcpkgs/intellij-idea-community-edition/template:
```diff
@@ -12,6 +12,7 @@ distfiles="https://download.jetbrains.com/idea/ideaIC-${version}-no-jbr.tar.gz"
 checksum=c2e4ac38b93349d0f9454720c58a5b7636b7e6cd04f9bb24008deed026350c8f
 repository=nonfree
 nopie=yes
+conflicts="intellij-idea-ultimate-edition"
```
Enough to set in one of packages and would need revision increment.

2asoft replied:
Will you be opposed to leaving it in both?
My intention here is to make both the xbps system aware of the conflict and also let whoever touches the templates know, and if one of the templates doesnt have the `conflicts`, it isnt clear there is a conflict until an install is attempted.

Chocimier replied:
Fine, leave it.

Chocimier commented on srcpkgs/CLion/template:
```diff
@@ -0,0 +1,64 @@
+	vcopy product-info.json ${TARGET_PATH}
+	vcopy build.txt ${TARGET_PATH}
+
+	ln -sf ${TARGET_PATH}/bin/clion.sh ${DESTDIR}/usr/bin/${pkgname}
```
This creates broken relative link.

Chocimier commented:
Close to ready. Some new versions came up, do you want to update before merging?

2asoft commented:
> Close to ready. Some new versions came up, do you want to update before merging?

Thanks for noticing, I saw that, was waiting to resolve all other items before upgrading.
Is this good to go now, or are there move improvements to be made here?

Chocimier commented:
Something like `vcopy license/* usr/share/licenses/${pkgname}` is more robust than looping over predefined directories.
Why there is no i686 builds anymore?

2asoft commented:
> Something like `vcopy license/* usr/share/licenses/${pkgname}` is more robust than looping over predefined directories.

`vcopy` is very nice, but the `custom` license requires `vlicense` calls (per the XLINT rules in Travis).

> Why there is no i686 builds anymore?

It's not an architecture I run, so I cant guarantee it works.
If a Void user ever needs this, I can work with them to add these and make sure it works.

Chocimier commented:
> custom license requires vlicense calls

Don't get xlint that strict. That rule was created to make people aware that license should be included, but if vcopy is used to include them it's okay to ignore error as false positive. Better to have simple template. Or one can silence error with comment containing word `vlicense`.

> I cant guarantee it works.

It may look irresponsibly, but I don't think anybody test all 9 builds for 5 architectures of regular build-from-source software. If one build is tested, all travis build passes and template is not too hackish, it is fine to ship.

2asoft commented:
> Don't get xlint that strict. That rule was created to make people aware that license should be included, but if vcopy is used to include them it's okay to ignore error as false positive. Better to have simple template. Or one can silence error with comment containing word `vlicense`.

Ok, done.

> If one build is tested, all travis build passes and template is not too hackish, it is fine to ship.

That's fair.
Added the i686 builds back.

Chocimier commented:
Pushed all but intellij.
For intellij, if only reason for conflict is /usr/bin/idea, it would be better to add to  idea-community and idea-ultimate, and `alternatives` on idea.

2asoft commented:
`/usr/bin/idea` is indeed the reason for `conflicts`.
I considered renaming, but that's a breaking change.
Good to know about `alternatives`, I think this solves the issue, thanks for pointing it out!

Updated the PR.

the-maldridge commented:
@Chocimier please seek additional review in the future from pkg-committers before accepting restricted packages.

Chocimier commented:
@the-maldridge: I will.

@2asoft: Changes on community needs revision increment and separate commit.

2asoft commented:
Increased the `intellij-idea-community-edition` revision and made it a separate commit.

Chocimier commented:
@2asoft Can you run this? For me it errors with message
```
JDK Required: JDK classes seem to be not on IDEA classpath.
Please ensure you run the IDE on JDK rather than JRE.
```

2asoft commented:
Interesting, it doesnt for me either.
I've run into this problem when I was running IDEA on a non-JetBrains JDK, but I didn't expect it here.
The Community edition also gives me the same error, by the way.

2asoft commented:
It has always been my intention to add the JetBrains JDK package to Void as well, to make sure their IDEs run on the best supported JDK.
I think I'll wait with this PR until I get that in place.

Chocimier commented:
@2asoft @shizonic: It works for me when `JAVA_HOME` is set to `/usr/lib/jvm/openjdk-1.8.0_202` as opposite to `/usr/lib/jvm/openjdk-1.8.0_202/jre`. If that fixes things for you too, it may be worth to add that tip to INSTALL.msg.

2asoft commented:
That would be a good tip to add to `INSTALL.msg`, but I think it shouldn't be necessary.

IMO it is wrong that JAVA_HOME is set to the /jre location in the first place (the `openjdk-jre` package does that).
If I'm spending time on this at all, I'll just get JetBrains' own JRE/JDK building and installable on Void.
Their builds have some performance patches, and the difference is noticeable in IntelliJ performance.

I just need to find the time to do it.

Chocimier commented:
@2asoft, please make another PR for JRE, this PR is big enough.

2asoft commented:
@Chocimier Good idea.
Here's the JDK PR #15629, will update this one once the JDK is merged.

Chocimier commented on srcpkgs/intellij-idea-ultimate-edition/template:
```diff
@@ -0,0 +1,72 @@
+revision=1
+archs="i686 x86_64"
+wrksrc="idea-IU-192.6817.14"
+depends="giflib libXtst jetbrains-jdk-11"
```
name of jdk changed

Chocimier commented on srcpkgs/jetbrains-jdk-11/template:
```diff
@@ -0,0 +1,39 @@
+# Template file for 'jetbrains-jdk-11'
```
please exclude from PR.

Chocimier commented:
@void-linux/pkg-committers: ok to merge #15629 then this restricted package?

Vaelatern commented:
I suppose so

the-maldridge commented:
I still dislike both, but I do not oppose the merge.

2asoft commented:
> I still dislike both, but I do not oppose the merge.

I'm curious to learn your reasons for disliking these, could you please share?

the-maldridge commented:
They're large, binary, restricted packages that could just as easily be downloaded and run from a home folder.  Sure there's convenience to having them in tree, but they're really not software that was designed to be maintained in this way.

2asoft commented:
Convenience was the main reason behind introducing these, that's correct.
I think letting the OS's package manager manage (un)installation of applications is a net benefit, even if the application is distributed as a large, restricted, binary package.
And while we are living in the days of effectively single-user machines, I believe this still applies.

Anyway, not trying to convince, only stating my point of view here.
Thank you for taking care of the Void packages system!

2asoft commented:
@whoever is going to review this: I'm done with all the force-pushing.

I originally included a line about installing `idea-ultimate` and `idea-community` as alternatives for `idea`.
However, ultimate would not start correctly when launched via the `idea` symlink.
I don't think the alternative is necessary, so I've removed it now. That's where all the pushing came from.

2asoft commented:
Ping. Would be nice to have this merged.
I know these arent exactly the kinds of packages that Void favors, but I wonder if there's something I could do better here to have these merged sooner and without annoying people with pings?

Vaelatern commented:
It's restricted. I can accept this. I intend to merge Sunday if I don't hear any objections.
